src/traces/densitymapbox/plot.js

```diff
@@ -0,0 +1,131 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var convert = require('./convert');
+
+function DensityMapbox(subplot, uid) {
+    this.subplot = subplot;
+    this.uid = uid;
+
+    this.sourceId = uid + '-source';
+
+    this.layerList = [
+        ['heatmap', uid + '-layer-heatmap']
+    ];
+
+    // previous 'below' value,
+    // need this to update it properly
+    this.below = null;
+}
+
+var proto = DensityMapbox.prototype;
+
+proto.update = function(calcTrace) {
+    var subplot = this.subplot;
+    var layerList = this.layerList;
+    var optsAll = convert(calcTrace);
+
+    subplot.map
+        .getSource(this.sourceId)
+        .setData(optsAll.geojson);
+
+    if(optsAll.below !== this.below) {
+        this._removeLayers();
+        this._addLayers(optsAll);
+    }
+
+    for(var i = 0; i < layerList.length; i++) {
+        var item = layerList[i];
+        var k = item[0];
+        var id = item[1];
+        var opts = optsAll[k];
+
+        subplot.setOptions(id, 'setLayoutProperty', opts.layout);
+
+        if(opts.layout.visibility === 'visible') {
+            subplot.setOptions(id, 'setPaintProperty', opts.paint);
+        }
+    }
+};
+
+proto._addLayers = function(optsAll) {
```

Comment:
Can we reuse the one from `src/traces/choroplethmapbox/plot.js`? I think it is identical

Reply:
We could, but I don't think it's worth it.